### PR TITLE
Implement the Mail System

### PR DESCRIPTION
**Pull request by 4Benj**
Changes:
- Get mail protos
- Send mail command
- Database support for mail and helper functions in Player class
- DatabaseHelper.GetAllPlayers (I'm not actually sure if this works and testing is required)

Things to be complete later:
- Mail template system (Saving a template to a database and send it using the `/sendmail` command. The `/sendmail` command already has a place for it to be easily implemented on [line 62](https://github.com/Grasscutters/Grasscutter/pull/115/files#diff-7acd01e31d9a50857d786a6c7b6e8b09be1833d0d7a264b2b1c6745fb7088c5bR62).
- Gift mail box. Currently it is completely disabled, it should be relatively simple to add by slightly modifying `GetAllMailRsp`. However, the amount of issues I have getting that working I'm kind of exhausting working on mail atm lol.

### CODE REVIEW

**Comment by BaiSugar** [comment]
try GetAllMailReq ,GetAllMailRsp,MailChangeNotify?

**Comment by lilmayofuksu** [comment]
image

It seems that server needs to reply to the GetAllMailReq packet

**Comment by 4Benj** [comment]
I've made a few updates and now the current issue is that the client seems to receive the mail and the mail icon is the menu has a badge indicating there is unread mail. However, there is nothing in the mail box when you check it...

It seems that it only updates when the client receives `MailChangeNotify` which doesn't make sense as that doesn't seem to be the same as the official servers. It's possible that the client receives `GetAllMailRsp` too early but if not I may have to send `MailChangeNotify` after sending `GetAllMailRsp`.

**Comment by PasserbyAlpha** [comment]
I've got some Exceptions when loading the game, and I guess it's because of the version of java (development branch has been switched to java 16 from 1.8). Add `@Entity` before the class definition of Mail, MailContent and MailItem will solve it (referring to https://github.com/Grasscutters/Grasscutter/issues/140#issuecomment-1107388851)

Besides, according to the message of MailChangeNotify, I doubt that maybe isRead and isAttachmentGot shouldn't appear in MailChangeNotify at all, instead of appearing with value false. At least those key don't appear in the message that I sniffed. I guess that client will only add MailChangeNotify without those key to original mail list. However, I'm not familiar with those libs and don't know how to make it. The only thing I learned is that we cannot just change boolean to Boolean and set null to them (or there will be exceptions)

**Comment by 4Benj** [comment]
> I've got some Exceptions when loading the game, and I guess it's because of the version of java (development branch has been switched to java 16 from 1.8). Add `@Entity` before the class definition of Mail, MailContent and MailItem will solve it (referring to https://github.com/Grasscutters/Grasscutter/issues/140#issuecomment-1107388851)

I kinda merged my project into the latest without realising that there going to huge changes like that and didn't actually test after merging because I was tired and thought it wasn't needed.

> Besides, according to the message of MailChangeNotify, I doubt that maybe isRead and isAttachmentGot shouldn't appear in MailChangeNotify at all, instead of appearing with value false. At least those key don't appear in the message that I sniffed. I guess that client will only add MailChangeNotify without those key to original mail list. However, I'm not familiar with those libs and don't know how to make it. The only thing I learned is that we cannot just change boolean to Boolean and set null to them (or there will be exceptions)

Thanks for that, I've been unable to sniff the packets myself for some particular reason, so I've had to reconstruct some things without a template and have just been interpreting some things based on their protos (which has worked out so far lol). When I get the chance I can remove isRead and isAttachmentGot from the packet being sent from the client later.

**Comment by 4Benj** [comment]
Big thanks to @lilmayofuksu for his help on the discord. 
This fucking game sometimes.
I had to send a "empty" packet notifying the game the game the mailbox was empty or it would fucking break.
I probably wouldn't of been able to do this without you so big thanks lol.

**Comment by 4Benj** [comment]
Now to add the missing packets and it'll be all done 😆 
See you after my next journey to the underworld :)

**Comment by 4Benj** [comment]
My madness is finally about to come to an end but I have some more testing to do but I'll do that tomorrow after school and after I get a bit of homework done (I've been putting it off for GC lol).
If anyone wants to test and encounter any issues, please post a comment here so I can fix it (or you can submit a PR to my repo)!

**Comment by 4Benj** [comment]
The mail system seems to be complete. There are further things that can be added later. However, the system works and the changes are minor and not required.

**Comment by 4Benj** [comment]
No matter what I add @KingRainbow44 always seems to have to review it :joy:
~~Apart from my last two~~

**Review by memetrollsXD — APPROVED** [review]

File: src/main/java/emu/grasscutter/database/DatabaseHelper.java
Comment (Konoyoru435) [inline src/main/java/emu/grasscutter/database/DatabaseHelper.java:94]: Thanks for this line of code, it helped me a lot, I was converting wetabq's kt to java, and there was a `toList`, and I found here that I could use `stream`

**Review by Konoyoru435 — APPROVED** [review]

File: src/main/java/emu/grasscutter/command/commands/SendMailCommand.java
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/command/commands/SendMailCommand.java:39]: update switch case to `case "foo" -> bar()` pls

File: src/main/java/emu/grasscutter/command/commands/SendMailCommand.java
Comment (KingRainbow44) [inline src/main/java/emu/grasscutter/command/commands/SendMailCommand.java]: private method maybe?

**Review by KingRainbow44 — APPROVED** [review]
looks good

**Comment by WetABQ** [comment]
after https://github.com/Grasscutters/Grasscutter/pull/226, u should del those Outer classes, which are generated by protoc
try to pull and merge, to resolve merge conflicts of u pr, thanks very much

**Comment by ghost** [comment]
Waiting on your conflict merge.

**Comment by 4Benj** [comment]
> Waiting on your conflict merge.

All done :)

**Comment by 4Benj** [comment]
Fixed the build issues and implemented @KingRainbow44 suggested changes

**Pull request merged**
